### PR TITLE
Fix: Publish types (+ validate and use types + generate type maps)

### LINKED ISSUE #15: Supporting typing
Hello! The team over at @stylelint is now using `support-hyperlinks` to show links in terminal output (see https://github.com/stylelint/stylelint/pull/5835). Currently, we've [manually defined types](https://github.com/stylelint/stylelint/pull/5835/files/fc6b53731abbd0f9969b34d8b6071e507e7f3f52#diff-1dfae1a47b7f5557a290b21c0459786cf8f44c25eba59339d3bcd916e1ee8213) for the package, though I think first-class/integrated support would be great!

Do you have a preferred way of doing this with `support-hyperlinks` (ex bundling it with the package, creating something under `@types/`, etc.)? Happy to contribute however I can!

### PR DESCRIPTION
**Pull request by voxpelli**
I noticed that this package was generating types but then told `npm` to only publish `browser.js` and `index.js` and not the actually generated types.

This fixes that + generates declarations maps + validates the types with TS `strict=true`.

It also adds a `tsconfig.json` to tell eg. VSCode that hey, this repository wants you to make use of TS types in `index.js`.

Tell me if you want me to pull something of this PR and I'll make it more granular 👍

_Note:_ I bumped the lowest supported Node-version to a [reasonable current LTS-focused](https://twitter.com/voxpelli/status/1521885437001932807) version range so that the `tsconfig.json` file could be pointed towards a reasonable fitting definition with [correct `lib` definition](https://github.com/tsconfig/bases/blob/96274973952a623d9f2e310987ceb4a19dcc362c/bases/node14.json#L6-L8): `es2020`

Fixes #15 

Closes #20

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -10,15 +10,17 @@
         "url": "github.com/jamestalmage"
     },
     "engines": {
-        "node": ">=8"
+        "node": "^14.18.0 || >=16.0.0"
```
Comment (sindresorhus) [inline package.json]:
```suggestion
        "node": ">=14.18"
```

File: index.js [hunk 2]
```diff
@@ -2,10 +2,14 @@
...
 
+/**
+ * @param {string} versionString
+ * @returns {{ major: number, minor: number, patch: number }}
+ */
 function parseVersion(versionString) {
 	if (/^\d{3,4}$/.test(versionString)) {
 		// Env var doesn't always use dots. example: 4601 => 46.1.0
-		const m = /(\d{1,2})(\d{2})/.exec(versionString);
+		const m = /(\d{1,2})(\d{2})/.exec(versionString) || [];
```
Comment (sindresorhus) [inline index.js:12]:
```suggestion
		const m = /(\d{1,2})(\d{2})/.exec(versionString) ?? [];
```

File: package.json [hunk 3]
```diff
@@ -10,15 +10,17 @@
...
         "prepublishOnly": "npm run create-types",
-        "test": "xo && nyc ava",
-        "create-types": "tsc index.js --allowJs  --declaration --emitDeclarationOnly"
+        "test": "xo && nyc ava && tsc",
+        "create-types": "tsc -p declaration.tsconfig.json"
     },
     "files": [
         "index.js",
+        "index.d.ts",
+        "index.d.ts.map",
```
Comment (sindresorhus) [inline package.json]: I don't think it should publish a Source Map

File: tsconfig.json [hunk 4]
```diff
@@ -0,0 +1,12 @@
+{
+	"extends": "@tsconfig/node14/tsconfig.json",
+	"files": [
+		"index.js"
+	],
+	"compilerOptions": {
+		"allowJs": true,
+		"checkJs": true,
+    "noEmit": true,
```
Comment (sindresorhus) [inline tsconfig.json]: Incorrect indentation

**Review by sindresorhus — CHANGES_REQUESTED** [review]

File: package.json (hunk 3 above)
Comment (voxpelli) [reply package.json]: Why not? It's just a way to ensure that "go to source" in eg. VSCode actually still goes to the source and not to the type file?

And its only this big:

```json
{"version":3,"file":"index.d.ts","sourceRoot":"","sources":["index.js"],"names":[],"mappings":"AA+BA,0CAHW;IAAE,KAAK,CAAC,EAAE,OAAO,GAAG,SAAS,CAAA;CAAE,GAC7B,OAAO,CAgFnB"}
```

File: tsconfig.json (hunk 4 above)
Comment (voxpelli) [inline tsconfig.json]:
```suggestion
		"noEmit": true,
```

File: package.json (hunk 3 above)
Comment (voxpelli) [reply package.json]: Related: https://github.com/Microsoft/TypeScript/issues/14479
And https://www.typescriptlang.org/docs/handbook/project-references.html#declarationmaps

But sure, I can remove it

Comment (voxpelli) [reply package.json]: Removed ✔️

File: declaration.tsconfig.json [hunk 5]
```diff
@@ -0,0 +1,9 @@
+
+{
+  "extends": "./tsconfig",
+  "compilerOptions": {
+    "declaration": true,
+    "noEmit": false,
+    "emitDeclarationOnly": true
+  }
+}
```
Comment (LitoMore) [inline declaration.tsconfig.json]:
```suggestion
{
	"extends": "./tsconfig",
	"compilerOptions": {
		"declaration": true,
		"noEmit": false,
		"emitDeclarationOnly": true
	}
}
```

Comment (voxpelli) [reply declaration.tsconfig.json]: Thanks 🙏 I wonder why my EditorConfig VSCode extension didn't catch / fix this 🤔

**Comment by voxpelli** [comment]
Anything I can do to help push this further?

File: package.json [hunk 6]
```diff
@@ -10,15 +10,16 @@
...
     "engines": {
-        "node": ">=8"
+        "node": ">=14.18"
     },
     "scripts": {
         "prepublishOnly": "npm run create-types",
-        "test": "xo && nyc ava",
-        "create-types": "tsc index.js --allowJs  --declaration --emitDeclarationOnly"
+        "test": "xo && nyc ava && tsc",
+        "create-types": "tsc -p declaration.tsconfig.json"
```
Comment (sindresorhus) [inline package.json]: Use the long-flag rather than `-p`.

File: index.js [hunk 7]
```diff
@@ -2,10 +2,14 @@
...
 
+/**
+@param {string} versionString
+@returns {{ major: number, minor: number, patch: number }}
+*/
 function parseVersion(versionString) {
 	if (/^\d{3,4}$/.test(versionString)) {
 		// Env var doesn't always use dots. example: 4601 => 46.1.0
-		const m = /(\d{1,2})(\d{2})/.exec(versionString);
+		const m = /(\d{1,2})(\d{2})/.exec(versionString) ?? [];
```
Comment (sindresorhus) [inline index.js]: You need to either upgrade XO or not use `??`. I suggest not using `??` as that's a less invasive change.

Comment (voxpelli) [reply index.js]: Fixed @sindresorhus

File: package.json (hunk 6 above)
Comment (voxpelli) [reply package.json]: Fixed

**Comment by sindresorhus** [comment]
https://github.com/jamestalmage/supports-hyperlinks/releases/tag/v3.0.0

**Comment by sindresorhus** [comment]
Not sure why master branch is failing when this PR passed.

https://github.com/jamestalmage/supports-hyperlinks/actions/runs/4363207180

**Pull request merged**